src/Merge/ExtraPackage.php

```diff
@@ -280,26 +282,61 @@ protected function mergeOrDefer(
             $this->logger->warning("Duplicate packages will be ignored.");
         }
 
-        $dups = array();
         foreach ($merge as $name => $link) {
-            if (isset($origin[$name]) && $state->ignoreDuplicateLinks()) {
-                $this->logger->info("Ignoring duplicate <comment>{$name}</comment>");
-                continue;
-            } elseif (!isset($origin[$name]) || $state->replaceDuplicateLinks()) {
-                $this->logger->info("Merging <comment>{$name}</comment>");
-                $origin[$name] = $link;
+            if (isset($origin[$name])) {
+                if ($state->ignoreDuplicateLinks()) {
+                    $this->logger->info("Ignoring duplicate <comment>{$name}</comment>");
+                    continue;
+                }
+
+                if ($state->replaceDuplicateLinks()) {
+                    $this->logger->info("Replacing <comment>{$name}</comment>");
+                    $origin[$name] = $link;
+                } else {
+                    $this->logger->info("Merging <comment>{$name}</comment>");
+                    $origin[$name] = $this->mergeConstraints($origin[$name], $link);
+                }
             } else {
-                // Defer to solver.
-                $this->logger->info(
-                    "Deferring duplicate <comment>{$name}</comment>"
-                );
-                $dups[] = $link;
+                $this->logger->info("Adding <comment>{$name}</comment>");
+                $origin[$name] = $link;
             }
         }
-        $state->addDuplicateLinks($type, $dups);
         return $origin;
     }
 
+    /**
+     * Merge package constraints.
+     *
+     * Adapted from Composer's UpdateCommand::appendConstraintToLink
+     *
+     * @param Link $origin The base package link.
+     * @param Link $merge  The related package link to merge.
+     * @return Link Merged link.
+     */
+    protected function mergeConstraints(Link $origin, Link $merge)
+    {
+        $parser = $this->versionParser;
+        $oldPrettyString = $origin->getConstraint()->getPrettyString();
+        $newPrettyString = $merge->getConstraint()->getPrettyString();
+        if (version_compare('2.0.0', PluginInterface::PLUGIN_API_VERSION, '>')) {
+            $constraintClass = 'Wikimedia\\Composer\\Merge\\MultiConstraint';
+        } else {
+            $constraintClass = 'Composer\\Semver\\Constraint\\MultiConstraint';
+        }
+        $newConstraint = $constraintClass::create(array(
+            $origin->getConstraint(),
+            $merge->getConstraint()
+        ), true);
+        $newConstraint->setPrettyString($oldPrettyString.', '.$newPrettyString);
```

Comment:
IMO to avoid ending up with a monstrosity of a pretty string here, which can lead to very confusing error output in case the solver can not resolve dependencies, I'd say if you are in composer2 please use https://github.com/composer/semver/blob/main/src/Intervals.php#L64 to check if the $origin is a subset of merge. If it is, then $merge can be dropped and you just return $origin IMO, and same for the opposite, as you really just want the smallest possible constraint of all here. i.e. if you want to merge `^2.3` and `^2.4`, as the latter is a subset of the former, you just want to keep `^2.4`, as it's the same as concatenating them into `^2.3, ^2.4`.

Reply:
Thanks. I'll implement these suggestions soon.